pxr/imaging/hdSt/materialXSyncDispatcher.cpp

```diff
@@ -0,0 +1,108 @@
+//
+// Copyright 2024 Pixar
+//
+// Licensed under the terms set forth in the LICENSE.txt file available at
+// https://openusd.org/license.
+//
+#include "pxr/imaging/hdSt/materialXSyncDispatcher.h"
+#include "pxr/imaging/hdSt/material.h"
+#include "pxr/imaging/hdSt/materialXFilter.h"
+
+#include "pxr/usd/sdf/path.h"
+
+namespace mx = MaterialX;
+
+PXR_NAMESPACE_OPEN_SCOPE
+
+HdSt_MaterialXSyncDispatcher::HdSt_MaterialXSyncDispatcher()
+{
+}
+
+HdSt_MaterialXSyncDispatcher::~HdSt_MaterialXSyncDispatcher()
+{
+}
+
+void
+HdSt_MaterialXSyncDispatcher::AddGeneratorTask(
+    std::unique_ptr<HdSt_MaterialXGeneratorTask> generatorTask,
+    HdStResourceRegistrySharedPtr& resourceRegistry)
+{
+    HD_TRACE_FUNCTION();
+
+
+    if ( resourceRegistry->ContainsMaterialXShader(
+        generatorTask->GetShaderHash() )) {
+        // We already have a shader for this topology.
+        return;
+    }
+
+    // Use a separate concurrent container for tracking unique generator tasks.
+    // The generated shader will be registered in the resource registry when
+    // the task completes.
+    auto insertResult =
```

Comment:
Another uniqueness check, this time between the different parallel tasks, using a dedicated data structure.